metrics_handler/metrics.py

```diff
@@ -0,0 +1,260 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import itertools
+import math
+
+from absl import logging
+import numpy as np
+import tensorflow as tf
+from tensorboard.backend.event_processing import event_multiplexer
+
+
+ALLOWED_COMPARISONS = ['greater', 'less', 'equal', 'less_or_equal']
+
+
+MetricPoint = collections.namedtuple('MetricPoint', ['metric_value', 'wall_time'])
+Threshold = collections.namedtuple('Threshold', ['type', 'value'])
+
+
+def read_metrics_from_events_dir(events_dir, tags_to_ignore=None):
+  """Collect the TensorBoard summary values for each metric.
+  
+  Args:
+    events_dir: path to location of TensorBoard summaries.
+    tags_to_ignore: set of TensorBoard tag names to skip.
+
+  Returns:
+    dict mapping TensorBoard tags to list of MetricPoint.
+  """
+  tags_to_ignore = tags_to_ignore or set()
+
+  em = event_multiplexer.EventMultiplexer()
+  em.AddRunsFromDirectory(events_dir)
+  em.Reload()
+
+  raw_metrics = collections.defaultdict(list)
+  for run, tags in em.Runs().items():
+    # 'Old-style' runs have a simple format and store values directly.
+    for tag in tags['scalars']:
+      if tag in tags_to_ignore:
+        continue
+      raw_metrics[tag].extend(
+          [MetricPoint(metric_value=x.value, wall_time=x.wall_time)
+          for x in em.Scalars(run, tag)])
+    # 'New-style' runs stores values inside of Tensor protos.
+    for tag in tags['tensors']:
+      if tag in tags_to_ignore:
+        continue
+      for t in em.Tensors(run, tag):
+        tensor_dtype = tf.dtypes.as_dtype(t.tensor_proto.dtype)
+        try:
+          val = np.frombuffer(
+              t.tensor_proto.tensor_content,
+              tensor_dtype.as_numpy_dtype).tolist()
+          assert len(val) == 1  # There should be 1 value per tensor.
+          raw_metrics[tag].append(
+              MetricPoint(metric_value=val[0], wall_time=t.wall_time))
+        except ValueError as e:
+          logging.warning(
+              'Unable to parse tag: `{}` from tensor_content: {}. '
+              'Error: {}. Consider adding this tag to tags_to_ignore '
+              'in config.'.format(tag, t.tensor_proto.tensor_content, e))
+
+  return raw_metrics
+
+
+def aggregate_metrics(raw_metrics, default_strategies, metric_strategies=None):
+  """Aggregate raw TensorBoard metrics according to collection config.
+
+  Available aggregation strategies: `final`, `min`, `max`.
+
+  Args:
+    raw_metrics: dict mapping TensorBoard tags to list of MetricPoint.
+    default_strategies: list of aggregation strategies to use as the default.
+    metric_strategies: dict mapping tags to aggregation strategies.
+
+  Returns:
+    dict mapping metric name to MetricPoint.
+  """
+  if not raw_metrics:
```

Comment:
I think it's better to return empty dict and log a warning for empty raw_metrics since not all models will write to Tensorboard.

If we have raw_metrics but no `default_strategies`, I could go either way on warning vs. error but probably would prefer logging an error and continuing the run. That second change can be later though

Reply:
Good idea. Done.